### PR TITLE
Fuzzyhashing

### PR DESCRIPTION
**Pull request by mcutshaw**
### Summary

If merged this pull request will address #149 

### Proposed changes

This request adds a fuzzy hash plugin using SSDEEP and TLSH. It places output in the metadata field of the main output BOM.

### CODE REVIEW

**Comment by nightlark** [comment]
Awesome! I'm out this upcoming week on vacation with limited internet access, so I might not be able to fully review the changes until March 11.

Initial thoughts are that overall it looks good, the main things are:
* pre-commit lint check about using a generator instead of `all` with a list comprehension (or restructure it to avoid creating list in order to perform the check)
* adding a brief docstring description for the fuzzy hashes hookimpl function
* the description in the pyproject.toml file looks like it may need updating
* maybe some error checks can be added to detect if there is an issue with the copy of `ssdeep` that was installed?

**Comment by mcutshaw** [comment]
> Awesome! I'm out this upcoming week on vacation with limited internet access, so I might not be able to fully review the changes until March 11.
> 
> Initial thoughts are that overall it looks good, the main things are:
> 
>     * pre-commit lint check about using a generator instead of `all` with a list comprehension (or restructure it to avoid creating list in order to perform the check)
> 
>     * adding a brief docstring description for the fuzzy hashes hookimpl function
> 
>     * the description in the pyproject.toml file looks like it may need updating
> 
>     * maybe some error checks can be added to detect if there is an issue with the copy of `ssdeep` that was installed?

The requested changes have been made. Also I made SSDEEP its own dependency group, and check whether we can import it. If we can't we just run with TLSH, hopefully that should simplify things.

**Comment by mcutshaw** [comment]
@nightlark Would you be able to review the most recent changes?

File: plugins/fuzzyhashes/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  libtool
+  build-essential
+  automake
+
+To install SSDEEP run the following:
+`pip install .[ssdeep]` or `pip install -e .[ssdeep]`
+
+## Uninstalling
+
+The plugin can be uninstalled with `pip uninstall surfactantplugin_fuzzyhashes.py`.
```
Comment (nightlark) [inline plugins/fuzzyhashes/README.md:39]:
```suggestion
The plugin can be uninstalled with `pip uninstall surfactantplugin-fuzzyhashes`.
```

File: plugins/fuzzyhashes/README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+```json
+{
+    "ssdeep": "3072:zk9IYDIW/+wxfiqV/jKneO1S4r88117lHc7ws47Fg5Q+ZLgFYY5:zsIYzpQqV/YLr8811P5",
+    "tlsh": "T1C3449303A267DC9FC4445AB105A75168FB38FC16CF36BB1BB242B73E6A31F009EA5640"
+}
+```
+
+Surfactant features for controlling which plugins are enabled/disabled can be used to control
+whether or not this plugin will run using the plugin name `surfactantplugin_fuzzyhashes.py` (the name given in
```
Comment (nightlark) [inline plugins/fuzzyhashes/README.md:22]:
```suggestion
whether or not this plugin will run using the plugin name `surfactantplugin_fuzzyhashes` (the name given in
```

File: plugins/fuzzyhashes/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+classifiers = [
+    "Programming Language :: Python :: 3",
+    "Environment :: Console",
+    "Operating System :: MacOS",
+    "Operating System :: Microsoft :: Windows",
+    "Operating System :: POSIX :: Linux",
+    "License :: OSI Approved :: MIT License",
+]
+dependencies = [
+    "python-tlsh",
```
Comment (nightlark) [inline plugins/fuzzyhashes/pyproject.toml:24]:
```suggestion
    "py-tlsh",
```

It looks like py-tlsh is the (slightly) more up to date package for TLSH: https://pypi.org/project/py-tlsh/

**Comment by nightlark** [comment]
The check for SSDEEP as an optional dependency looks like a good option -- I opened an issue on the python-ssdeep repository [about pre-compiled binary wheels](https://github.com/DinoTools/python-ssdeep/issues/67) for common platforms, and checking on the maintenance status. It seems like the python-ssdeep package likely doesn't have an active maintainer at the moment.

**Review by nightlark — APPROVED** [review]
I had some issues with ssdeep, which I think are related to the ssdeep python package itself (build errors, returned string from ssdeep.hash a string in the form `b'<hash>'` which looks like it is returning bytes turned into a str instead of just a str).

**Pull request merged**